[4.0] Subform custom element

Fedik opened the pull request:
### Summary of Changes
Convert subform.js to custom element `<joomla-field-subform />`


### Testing Instructions
Add subform field in to an any form, and check whether it works as usual.

### Documentation Changes Required
yes I think


### Todo:

- [x]  make sorting to work (important)
- [x]  nested forms (optional)


@dgt41 Christmas present to you 😄

Fedik committed:
subform-custom-element, basic add/remove for rows

Fedik committed:
subform-custom-element, fixUniqueAttributes and events

Fedik committed:
subform-custom-element: fix table row rendering

Fedik committed:
subform-custom-element: test sortable with shopify/draggable

dgrammatiko commented:
Great work @Fedik Merry Christmas to you too!! 🎉

dgrammatiko committed:
make an es5 copy as well

This PR just moves the source code to the /build/webcomponents folder so the command `$ grunt createElements` will take care of the es5 (for IE11)

dgrammatiko commented:
@Fedik check https://github.com/Fedik/joomla-cms/pull/4

dgrammatiko commented:
Another thing about the sortable functionality: it needs to be keyboard accessible. I have no clue if either the draggula or the airbnb script are accessible! Anyways some time ago @brianteeman pointed out some resource about accessible drag and drop: https://www.sitepoint.com/accessible-drag-drop/
(I think it's the correct one, but I couldn't find the actual discussion).
All and all we need to re-evaluate the script for drag and drop considering accessibility!

Fedik committed:
Merge pull request #4 from dgt41/subform-custom-element

Make an es5 copy as well

Fedik commented:
@dgt41 current problem is to make the sorting to work, at least  😄 

because it is now a custom element, I do not think it will be good idea to have "external dependency"

as you already told, can be need to merge in to single file

Fedik commented:
@dgt41 and idea,
what if we drop "drag sorting" and just make "old school" a two arrows 'move Up'/'move Down' 😄 
it is perfect for accessibility and should not be hard to implement :neckbeard:

dgrammatiko commented:
@Fedik let's create couple codepens with deferent approaches and let the a11y and design team decide which way the want to go. For me anything that fulfils the requirements is good enough! 😃

C-Lodder commented:
@Fedik would be easy to implement but we'd be going backwards rather than forwards. J3 supports drag-n-drop and J4 only allows you to move rows 1 place at a time

dgrammatiko commented:
@C-Lodder J3's drag and drop is **not** accessible, the point is accessibility here, if we can replicate J3 dnd in accessible fashion lets do it, else let's do something that is accessible...

C-Lodder commented:
Brian shared a link for an a11y dnd a while back. It's definitely possible

Fedik commented:
>  but we'd be going backwards rather than forwards

it an evolution loop 😄 
well, just an idea 😉 

for now I do not have any solution

> Brian shared a link for an a11y dnd a while back. It's definitely possible

the problem with that example, that it use HTML5 `draggable` which do not work good on mobiles

brianteeman commented:
up - down arrows are not an improvement and they are not better for accessibility

It is perfectly possible to provide keyboard support for drag and drop
See https://www.sitepoint.com/accessible-drag-drop/ for info
http://jspro.brothercake.com/multidrag/demo4b.html for demo

dgrammatiko commented:
@Fedik do you want to experiment on the code of site point that @brianteeman posted above?
I could spend some time to make this behave as the sortable

Fedik commented:
@dgt41 I will try on next weekend, 
but any help are welcome 😉

Fedik committed:
Use of HTML Drag and Drop API

Fedik committed:
handle dragend

Fedik committed:
mouse/keyboards selection

dgrammatiko commented:
@Fedik check
- https://github.com/RubaXa/Sortable
- https://codepen.io/taoofbean/pen/rnCvK

I can live with a CE wrapper (joomla-sortable) for this lib ;)

Fedik commented:
@dgt41  I seen that, but somehow I do not liked that 😄 

If use of external lib, then I would stop on https://github.com/Shopify/draggable they declare Accessibility support https://github.com/Shopify/draggable/search?q=accessibility&type=Issues&utf8=%E2%9C%93

for now I try native HTML 5 dragable API, with some polyfill for mobile browsers

Fedik commented:
note for me

keymap:

`tab` to switch element / switch drop target
`space` (or modifier + `space`) to select
`ctrl+m` / `enter` to place
`esc` to cancel

Fedik committed:
Sync. Merge branch '4.0-dev' into subform-custom-element

Conflicts:
	layouts/joomla/form/field/subform/repeatable-table.php
	layouts/joomla/form/field/subform/repeatable.php

Fedik committed:
make "drag sorting" to work, with respect a move handler

Fedik committed:
use `dragenter` event to show actual position of the moved row

Fedik committed:
Implement keyboard sorting

Fedik committed:
Simplify markup, add needed attributes via script

Fedik commented:
for now,
Desktop drag sorting, works.

Keyboard sorting, works:

1. `tab` to navigate to needed row,
2. (ctr,alt,shift) + `space` select the row,
3. `tab` to select destination,
4. `enter` to place selected row in to destination

`esc` to cancel selection

Fedik committed:
Clear the item

Quy commented on build/webcomponents/js/field-subform/field-subform.js:
```diff
@@ -0,0 +1,298 @@
+                this.containerWithRows.append(row);
+            }
+
+            //add marker that it is new
```
coding style

Fedik replied:
that is last problem here for now 😉

Quy commented on build/webcomponents/js/field-subform/field-subform.js:
```diff
@@ -0,0 +1,298 @@
+
+            //add marker that it is new
+            row.setAttribute('data-new', '1');
+            // fix names and id`s, and reset values
```
capitalize `f`. change `id's` to `ids` or `IDs`

dgrammatiko commented:
Great! One thing though the source in in `build/webcomponents/...` not directly in the `media/system/webcomponents`. The later is the compiled/transpiled version

Fedik commented:
I still need to add keyboard navigation to add/remove buttons.
Test the mobile `dragable` polyfil.
And try something with "nested" forms, if I will have a mood 😄

dgrammatiko commented on media/system/webcomponents/js/joomla-field-subform.js:
```diff
@@ -265,6 +262,54 @@
+            // dragstart event to initiate mouse dragging
+            this.addEventListener('dragstart', function(event) {
+                // Make sure we handle correct children
+                if (closest(event.target, 'joomla-field-subform') !== that) {
```
Can also be: `if (!that.contains(event.target))`

Fedik replied:
need to check

Fedik replied:
no, contains() will do different, 
and I need to be sure that the parent is correct, to handle `nested`  sorting in future

dgrammatiko replied:
do we support multiple levels of subforms? wow inception mode on

Fedik replied:
no for now, but I thought about it, and made some preparation already

dgrammatiko commented on media/system/webcomponents/js/joomla-field-subform.js:
```diff
@@ -265,6 +262,54 @@
+
+                // Make sure the target is/in the correct container
+                if (event.target !== that.containerWithRows
+                    && (that.rowsContainer && closest(event.target, that.rowsContainer) !== that.containerWithRows)) {
```
Same here: `that.containerWithRows.contains(event.target)`

Fedik committed:
Move to build/webcomponents/

Fedik committed:
Keybnav support for a row add/remove actions

Fedik committed:
update es5

Fedik committed:
Use of constructor() instead of connectedCallback() to prevent double events binding

Fedik committed:
Implement touch interaction

Fedik committed:
"Switch position" helper

Fedik committed:
Add basic style to hinglight [aria-grabbed="true"]

Fedik committed:
update es5

Fedik commented:
@dgt41 can you please have a look, I cannot make es6 minyfied, there some random error 😄 
![screen 2018-01-08 15 48 15 751x206](https://user-images.githubusercontent.com/1568198/34673622-89beee0a-f48b-11e7-8b53-5163d1d36bbc.png)

Fedik committed:
Filter out nested

dgrammatiko commented:
@Fedik run
```
npm install git://github.com/gruntjs/grunt-contrib-uglify.git#harmony --save-dev
```
and change in the Gruntfile.js the line
```js
grunt.loadNpmTasks('grunt-contrib-uglify');
```
to 
```js
grunt.loadNpmTasks('grunt-contrib-uglify-es');
```

then run again ```grunt createElement```

Fedik committed:
update minified version

Fedik commented:
@dgt41 works. thanks!

Fedik committed:
Remove old subform script

Fedik committed:
Basic nested support

Fedik commented:
Now it support a base nesting, however I would not recommend to use a complex nested forms, and complex fields in a nested forms. example an editor field should not be in nested.

Fedik commented:
### This is ready for testing.
Use any existing subform field, and make sure it still work.
Or use this example:

```xml
<field name="field-name" type="subform" multiple="true" label="Subform Field">
	<form>
		<field type="text" name="text" label="text" />
	</form>
</field>
```

For nested:
```xml
<field name="field-name-other" type="subform" multiple="true" label="Subform Field">
    <form>
        <field type="text" name="text" label="text" />
        <field name="subform-nested" type="subform" multiple="true" label="Nested Field"
            layout="joomla.form.field.subform.repeatable-table">
            <form>
                <field type="text" name="text-nested" label="text2" />
            </form>
        </field>
    </form>
</field>
```

brianteeman commented:
@Fedik if you can resolve the conflicts I will give it a try

Fedik committed:
Sync. Merge branch '4.0-dev' into subform-custom-element

Conflicts:
	administrator/templates/atum/css/template-rtl.min.css
	administrator/templates/atum/css/template.min.css
	grunt-settings.yaml
	media/system/js/fields/subform-repeatable.min.js
	media/system/webcomponents/js/joomla-field-user-es5.min.js

Fedik committed:
update template after conflict

Fedik commented:
@brianteeman done

brianteeman commented:
Thanks will test tomorrow

brianteeman commented:
Tested with a subform and cannot work out how to get the sorting by keyboard to work

Quy commented on layouts/joomla/form/field/subform/repeatable-table.php:
```diff
@@ -72,12 +72,11 @@
-			data-repeatable-element="tr.subform-repeatable-group"
-			data-rows-container="tbody" data-minimum="<?php echo $min; ?>" data-maximum="<?php echo $max; ?>">
+		<joomla-field-subform class="subform-repeatable" name="<?php echo $name; ?>"
+			button-add=".group-add" button-remove=".group-remove" button-move="<?php echo empty($buttons['move']) ? '' : '.group-move' ?>"
```
Add semicolon after `'.group-move'`

Quy commented on layouts/joomla/form/field/subform/repeatable-table.php:
```diff
@@ -103,10 +103,10 @@
-		<?php echo $this->sublayout($sublayout, array('form' => $tmpl, 'basegroup' => $fieldname, 'group' => $fieldname . 'X', 'buttons' => $buttons)); ?>
-		</script>
+		<template class="subform-repeatable-template-section" style="display: none;"><?php
+			echo trim($this->sublayout($sublayout, array('form' => $tmpl, 'basegroup' => $fieldname, 'group' => $fieldname . 'X', 'buttons' => $buttons)));
```
Make same line `<?php echo trim($this->sublayout($sublayout, array('form' => $tmpl, 'basegroup' => $fieldname, 'group' => $fieldname . 'X', 'buttons' => $buttons))); ?>`

Fedik replied:
it will stay as is, it is needed to avoid useless tabs, a new lines and a spaces inside the template

Fedik committed:
update style for aria-grabbed

Fedik commented:
> Tested with a subform and cannot work out how to get the sorting by keyboard to work

@brianteeman I knew it will be hard, and it is really hard :neckbeard: 

Once again for Keyboard sorting:

1. `tab` to navigate to needed row,
2. (ctr,meta,shift) + `space` select the row,
3. `tab` to select a destination,
4. `enter` to place selected row in to destination

### The sorting by keyboard in pictures

I have painted a couple pictures.

1. `tab` to navigate to needed row (I would recommend to click the label first, to get there faster), you should see "row in focus" (style different from OS & Browser): 
![screen 2018-01-20 17 26 05 871x417](https://user-images.githubusercontent.com/1568198/35185134-eb33835a-fe07-11e7-9417-1c7885a66f21.png)

2. `ctrl` + `space` OR `shift` + `space` OR `meta` + `space`  to select the row (aria-grabbed="true"), you should notice darkblue shadow:
![screen 2018-01-20 17 26 18 871x417](https://user-images.githubusercontent.com/1568198/35185148-3b6de824-fe08-11e7-9c33-8e018ba7bd22.png)

3. now use `tab` to focus the destination, where selected row should be placed:
![screen 2018-01-20 17 26 40 871x417](https://user-images.githubusercontent.com/1568198/35185178-ce980814-fe08-11e7-8d93-bd8dcb630ad0.png)

4. push `enter`, now the selected row should be placed at destination:
![screen 2018-01-20 17 27 01 871x417](https://user-images.githubusercontent.com/1568198/35185194-f73e1ca4-fe08-11e7-8da0-1062fec89463.png)

5. Done!

But it is really hard, especial if you have a lot of fields inside, you should push `tab` 1000 times.
I tried to follow a recommendations from https://www.sitepoint.com/accessible-drag-drop/ and some other sources.

Any suggestions to improve are welcome 😉

Fedik committed:
Sync. Merge branch '4.0-dev' into subform-custom-element

 Conflicts:
	administrator/templates/atum/css/template-rtl.min.css
	administrator/templates/atum/css/template.min.css

Fedik committed:
Update minified version for css, after conflict

brianteeman commented:
Thanks for the update. I will try again soon. I was definitely trying to do what you showed in the images.

Fedik commented:
>  I was definitely trying to do what you showed in the images.

Also there was problem with minimized template.min.css, and "select' action was invisible. 
Maybe this was a reason.
I have updated template.min.css in last commit

brianteeman commented:
ok it works now. ;)

One question.
After selecting a row I would expect the next tab would take me to the next row where I want to place the selected row. However i have to tab through all the elements in the row to get to the next row

Fedik commented:
> After selecting a row I would expect the next tab would take me to the next row where I want to place the selected row. However i have to tab through all the elements in the row to get to the next row

I would expect it also.

The problem that the rows use `tabindex="0"` and so "flow"  with other elements/input on the page.
I tried to set it `tabindex="1"` then it works a bit better, however I have read recommendation to do not use `tabindex` bigger than zero, because it can confuse screen-reader, or something.

So for now it like this. 
Maybe in future we will get a better idea how to fix it.

dgrammatiko commented:
@wilsonge can you please merge this one?
@Fedik please fix the conflict

Fedik committed:
Merge branch '4.0-dev' into subform-custom-element

Conflicts:
	administrator/templates/atum/css/template-rtl.min.css
	administrator/templates/atum/css/template.min.css
	media/system/webcomponents/js/joomla-field-user-es5.min.js

Fedik committed:
update template.min.css after conflict

Fedik commented:
done

frogydiak commented:
How did you guys tested this subform field? The ability to move cloned fields is not working in my test. Someone told me in my other post that jqueryUI has been deprecated in J!4 but I'm getting error with it in the subform field. Here's the screen capture:

http://take.ms/eNrxb

mbabker commented:
> Someone told me in my other post that jqueryUI has been deprecated in J!4 but I'm getting error with it in the subform field

Not all deprecated items have been removed yet.  If it's marked deprecated in 3.x and still exists today in the 4.0 code, that doesn't necessarily mean it's going to survive to the stable release.  Remember 4.0 is very much still a work in progress.

dgrammatiko commented:
@frogydiak well jquery-ui is not used in this PR.
Go to issues.joomla.org and click on the link   `Fedik:subform-custom-element` then install Joomla...
<img width="1196" alt="screen shot 2018-02-06 at 21 11 45" src="https://user-images.githubusercontent.com/3889375/35881868-86b6b66c-0b82-11e8-80d9-fffce2cbbdde.png">

frogydiak commented:
@mbabker Thanks for the explanation, I'm quite new to github joomla and wanted to participate on reporting issues. I am **confused** to say the least.  LOL! If you look at dgt41 saying above, he's pointing me to another website that is loading this thread. What?! Do I have to know in-depth  Joomla developer jargons to be able to help out? Why do you guys decided to have another site for issue tracker when you can just redirect to github? Another layer of confusion and system to develop and manage. There's no clear guidelines as what to expect in J!4 branch. That explanation you did could have been in the README.md but it seems to me it was just copied from previous versions. *scratching my head... confused!*

dgrammatiko committed:
tests template

Fedik committed:
Make a subform table layout to be more strict to rows container

Fedik committed:
Sync. Merge branch '4.0-dev' into subform-custom-element

Fedik committed:
Merge branch 'pull5' @dgt41 subform-custom-element-tests into subform-custom-element

 Conflicts:
	karma.conf.js
	tests/javascript/validate/spec.js

Fedik committed:
Update tests

dgrammatiko commented:
@Fedik can you resolve the conflict here?

Fedik committed:
Sync. Merge branch '4.0-dev' into subform-custom-element

 Conflicts:
	karma.conf.js

Fedik commented:
done

Commit:
Update repeatable-table.php

Commit:
Update repeatable.php

dgrammatiko commented:
@Fedik can you merge: https://github.com/Fedik/joomla-cms/pull/7 and https://github.com/Fedik/joomla-cms/pull/6

Then I can mark my test as successful here

Fedik committed:
Merge pull request #6 from dgrammatiko/patch-3

Update repeatable-table.php

Fedik committed:
Merge pull request #7 from dgrammatiko/patch-4

Update repeatable.php

Fedik commented:
@dgrammatiko done, thanks!

dgrammatiko commented:
I have tested this item :white_check_mark: successfully on bdaf42da91988b1c8d7ac1ca9ebc2d946ce22ef7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19184">issues.joomla.org/tracker/joomla-cms/19184</a>.</sub>

laoneo commented:
As the tests got moved to a separate repo you need to create a pr with the tests in the [JS testing repo](https://github.com/joomla/test-javascript) and remove them from here.

brianteeman commented:
@laoneo and thats why it was not the best decision

dgrammatiko commented:
@laoneo since all tests are basically nullified we can safely remove them in this PR as well.
Of course at some later point (eg API is stable) we will spent some time redoing them, but right now I guess this is not a show stopper

dgrammatiko commented:
> and thats why it was not the best decision

so I will agree this is bad for the short term **but** once the tests are done correctly this would be a better solution. (the main advantage is that if a PR breaks some tests this will be very obvious and since we want to keep B/C tests should never break, at least that's part of the theory there)

laoneo commented:
@brianteeman I wasn't part of that decision and do not know all the pros and cons. For now, if the tests are unnecessary then lets remove them in this pr and add them later.

dgrammatiko commented:
@Fedik can you resync this (basically you will have to delete the tests as these were deleted few days ago)

Fedik committed:
Sync. Merge branch '4.0-dev' into subform-custom-element

 Conflicts:
	tests/javascript/joomla-field-subform/fixtures/fixture.html
	tests/javascript/test-main.js

Fedik committed:
Remove joomla-field-subform tests

Fedik committed:
Merge branch 'subform-custom-element' of github.com:Fedik/joomla-cms into subform-custom-element

laoneo commented:
@Fedik can this be tested by creating a new list custom field as it uses subforms?

Fedik commented:
@laoneo it should work as it was before, everywhere where `subform` field was used :wink: 
so yes

laoneo committed:
Merge branch '4.0-dev' into subform-custom-element

laoneo commented:
I have tested this item :white_check_mark: successfully on bdaf42da91988b1c8d7ac1ca9ebc2d946ce22ef7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19184">issues.joomla.org/tracker/joomla-cms/19184</a>.</sub>

laoneo committed:
Merge branch '4.0-dev' into subform-custom-element

dgrammatiko commented:
I have tested this item :white_check_mark: successfully on bdaf42da91988b1c8d7ac1ca9ebc2d946ce22ef7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19184">issues.joomla.org/tracker/joomla-cms/19184</a>.</sub>

ghost commented:
Ready to Commit after two successful tests.

phalouvas commented:
Below problem occurred after this update. I tried to see where exactly is the problem, but failed.  The test was on Microsoft Edge browser.
![form_problem](https://user-images.githubusercontent.com/5654203/39506539-dee88562-4de1-11e8-86e9-2682ac35da45.png)

laoneo commented:
Can you please open a new issue with the problem. Thanks for understanding.
